lib/symbol-provider.js

```diff
@@ -0,0 +1,371 @@
+  }
+
+  addConfigEntry (config) {
+    for (let type in config) {
+      let options = config[type]
+      if (this.config[type] == null) { this.config[type] = {} }
+      if (options.selector != null) { this.config[type].selectors = Selector.create(options.selector) }
+      this.config[type].typePriority = options.typePriority != null ? options.typePriority : 1
+      this.config[type].wordRegex = this.wordRegex
+
+      let suggestions = this.sanitizeSuggestionsFromConfig(options.suggestions, type)
+      if ((suggestions != null) && suggestions.length) { this.config[type].suggestions = suggestions }
+    }
+  }
+
+  sanitizeSuggestionsFromConfig (suggestions, type) {
+    if ((suggestions != null) && Array.isArray(suggestions)) {
+      let sanitizedSuggestions = []
+      for (let i = 0; i < suggestions.length; i++) {
+        let suggestion = suggestions[i]
+        if (typeof suggestion === 'string') {
+          sanitizedSuggestions.push({text: suggestion, type})
+        } else if (typeof suggestions[0] === 'object' && ((suggestion.text != null) || (suggestion.snippet != null))) {
+          suggestion = _.clone(suggestion)
+          if (suggestion.type == null) { suggestion.type = type }
+          sanitizedSuggestions.push(suggestion)
+        }
+      }
+      return sanitizedSuggestions
+    } else {
+      return null
+    }
+  }
+
+  uniqueFilter (completion) { return completion.text }
+
+  paneItemIsValid (paneItem) {
+    // TODO: remove conditional when `isTextEditor` is shipped.
+    if (typeof atom.workspace.isTextEditor === 'function') {
+      return atom.workspace.isTextEditor(paneItem)
+    } else {
+      if (paneItem == null) { return false }
+      // Should we disqualify TextEditors with the Grammar text.plain.null-grammar?
+      return (paneItem.getText != null)
+    }
+  }
+
+  /*
+  Section: Suggesting Completions
+  */
+
+  getSuggestions (options) {
+    if (!options.prefix) {
+      return
+    }
+
+    if (options.prefix.trim().length < this.minimumWordLength) {
+      return
+    }
+
+    this.buildConfigIfScopeChanged(options)
+    let editor = options.editor
+    let bufferPosition = options.bufferPosition
+    let prefix = options.prefix
+
+    let numberOfWordsMatchingPrefix = 1
+    let wordUnderCursor = this.wordAtBufferPosition(editor, bufferPosition)
+    let iterable = editor.getCursors()
+    for (let i = 0; i < iterable.length; i++) {
+      let cursor = iterable[i]
+      if (cursor === editor.getLastCursor()) { continue }
+      let word = this.wordAtBufferPosition(editor, cursor.getBufferPosition())
+      if (word === wordUnderCursor) { numberOfWordsMatchingPrefix += 1 }
+    }
+
+    let buffers = this.includeCompletionsFromAllBuffers ? null : [this.editor.getBuffer()]
+    let symbolList = this.symbolStore.symbolsForConfig(
+      this.config, buffers,
+      prefix, wordUnderCursor,
+      bufferPosition.row,
+      numberOfWordsMatchingPrefix
+    )
```

Comment:
Is there a reason for this somewhat unorthodox formatting? This should be broken on every comma, right? Alternatively break once, and leave all parameters on a single line.